Feat/advanced fees

saledjenic opened the pull request:
This PR introduces a transaction settings option where the user should be able to select among 3 different options for the max fee (normal, fast, urgent) as well as fully customizing the details (base fee, priority fee, gas limit, nonce) for each transaction.

Issues that are covered by this PR:
- https://github.com/status-im/status-desktop/issues/16193
- https://github.com/status-im/status-desktop/issues/16194

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 79e9c9ad | [#1](https://ci.status.im/job/status-desktop/job/prs/job/macos/job/aarch64/job/package/job/PR-17118/1/) | 2025-01-23 12:44:05 | ~7 min | `macos/aarch64` | [:apple:`dmg`](https://status-im-desktop-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Desktop-250123-123656-79e9c9-pr17118-aarch64.dmg) |

virginiabalducci commented:
I started testing and I'll edit this comment as I go if I find any issues.

### Issue 1 
After changing Custom fees to an invalid amount, the default custom values are not reset and the fees do not load.
To reload the fees, the user has to cancel the Send transaction and start over.

@saledjenic is this related to what you said about hard-coded values for the estimated fees?

<details><summary>Details</summary>
<p>

https://github.com/user-attachments/assets/7315d2e7-1652-4f34-b661-e2aabbc2c004

https://github.com/user-attachments/assets/722438d6-5c1c-4b22-95cf-45b517601109


`WRN 2025-02-14 14:00:50.461Z qt warning                                 topics="qt" tid=28733564 text="Qt Quick Layouts: Polish loop detected. Aborting after two iterations." file=:0 category=default
WRN 2025-02-14 14:00:50.461Z qt warning                                 topics="qt" tid=28733564 text="QML StatusScrollView: Binding loop detected for property \"contentHeight\"" file=qrc:/app/AppLayouts/Wallet/popups/SignTransactionModalBase.qml:142 category=default
ERR 2025-02-14 14:00:51.567Z qt error                                   topics="qt" tid=28733564 text="\ndiv (qrc:/StatusQ/Core/Utils/AmountsArithmetic.qml:128)\nweiToGWei (qrc:/imports/utils/Utils.qml:995)\nexpression for text (qrc:/app/AppLayouts/Wallet/views/TransactionSettings.qml:280)\nonClicked (qrc:/app/AppLayouts/Wallet/popups/simpleSend/SendSignModal.qml:228)" file=qrc:/StatusQ/Core/Utils/AmountsArithmetic.qml:128 category=default
ERR 2025-02-14 14:00:51.568Z qt error                                   topics="qt" tid=28733564 text="\ndiv (qrc:/StatusQ/Core/Utils/AmountsArithmetic.qml:128)\nweiToGWei (qrc:/imports/utils/Utils.qml:995)\nexpression for text (qrc:/app/AppLayouts/Wallet/views/TransactionSettings.qml:244)\nonClicked (qrc:/app/AppLayouts/Wallet/popups/simpleSend/SendSignModal.qml:228)" file=qrc:/StatusQ/Core/Utils/AmountsArithmetic.qml:128 category=default
WRN 2025-02-14 14:00:51.578Z qt warning                                 topics="qt" tid=28733564 text="TypeError: Property 'div' of object NaN is not a function" file=qrc:/StatusQ/Core/Utils/AmountsArithmetic.qml:130 category=default
WRN 2025-02-14 14:00:51.578Z qt warning                                 topics="qt" tid=28733564 text="TypeError: Property 'div' of object NaN is not a function" file=qrc:/StatusQ/Core/Utils/AmountsArithmetic.qml:130 category=default
WRN 2025-02-14 14:00:55.016Z qt warning                                 topics="qt" tid=28733564 text="TypeError: Value is null and could not be converted to an object" file=qrc:/app/mainui/SendModalHandler.qml:596 category=default
WRN 2025-02-14 14:00:55.016Z qt warning                                 topics="qt" tid=28733564 text="TypeError: Value is null and could not be converted to an object" file=qrc:/app/mainui/SendModalHandler.qml:596 category=default
WRN 2025-02-14 14:00:55.016Z qt warning                                 topics="qt" tid=28733564 text="TypeError: Value is null and could not be converted to an object" file=qrc:/app/mainui/SendModalHandler.qml:596 category=default
WRN 2025-02-14 14:00:55.016Z qt warning                                 topics="qt" tid=28733564 text="TypeError: Cannot call method 'refreshTxSettings' of null" file=qrc:/app/AppLayouts/Wallet/popups/simpleSend/SendSignModal.qml:122 category=default
ERR 2025-02-14 14:01:02.407Z qt error                                   topics="qt" tid=28733564 text="\ndiv (qrc:/StatusQ/Core/Utils/AmountsArithmetic.qml:128)\nweiToGWei (qrc:/imports/utils/Utils.qml:995)\nupdateCustomFields (qrc:/app/AppLayouts/Wallet/popups/simpleSend/SendSignModal.qml:267)\nonSelectedFeeModeChanged (qrc:/app/AppLayouts/Wallet/popups/simpleSend/SendSignModal.qml:309)\nonClicked (qrc:/app/AppLayouts/Wallet/views/TransactionSettings.qml:210)\nonClicked (qrc:/StatusQ/Controls/StatusFeeOption.qml:168)" file=qrc:/StatusQ/Core/Utils/AmountsArithmetic.qml:128 category=default
WRN 2025-02-14 14:01:02.408Z qt warning                                 topics="qt" tid=28733564 text="TypeError: Property 'div' of object NaN is not a function" file=qrc:/StatusQ/Core/Utils/AmountsArithmetic.qml:130 category=default
ERR 2025-02-14 14:01:08.444Z qt error                                   topics="qt" tid=28733564 text="\ndiv (qrc:/StatusQ/Core/Utils/AmountsArithmetic.qml:128)\nweiToGWei (qrc:/imports/utils/Utils.qml:995)\nexpression for text (qrc:/app/AppLayouts/Wallet/views/TransactionSettings.qml:280)\nonClicked (qrc:/app/AppLayouts/Wallet/popups/simpleSend/SendSignModal.qml:228)" file=qrc:/StatusQ/Core/Utils/AmountsArithmetic.qml:128 category=default
ERR 2025-02-14 14:01:08.445Z qt error                                   topics="qt" tid=28733564 text="\ndiv (qrc:/StatusQ/Core/Utils/AmountsArithmetic.qml:128)\nweiToGWei (qrc:/imports/utils/Utils.qml:995)\nexpression for text (qrc:/app/AppLayouts/Wallet/views/TransactionSettings.qml:244)\nonClicked (qrc:/app/AppLayouts/Wallet/popups/simpleSend/SendSignModal.qml:228)" file=qrc:/StatusQ/Core/Utils/AmountsArithmetic.qml:128 category=default
WRN 2025-02-14 14:01:08.455Z qt warning                                 topics="qt" tid=28733564 text="TypeError: Property 'div' of object NaN is not a function" file=qrc:/StatusQ/Core/Utils/AmountsArithmetic.qml:130 category=default
WRN 2025-02-14 14:01:08.455Z qt warning                                 topics="qt" tid=28733564 text="TypeError: Property 'div' of object NaN is not a function" file=qrc:/StatusQ/Core/Utils/AmountsArithmetic.qml:130 category=default
WRN 2025-02-14 14:01:10.156Z qt warning                                 topics="qt" tid=28733564 text="TypeError: Value is null and could not be converted to an object" file=qrc:/app/mainui/SendModalHandler.qml:596 category=default
WRN 2025-02-14 14:01:10.156Z qt warning                                 topics="qt" tid=28733564 text="TypeError: Value is null and could not be converted to an object" file=qrc:/app/mainui/SendModalHandler.qml:596 category=default
WRN 2025-02-14 14:01:10.156Z qt warning                                 topics="qt" tid=28733564 text="TypeError: Value is null and could not be converted to an object" file=qrc:/app/mainui/SendModalHandler.qml:596 category=default
WRN 2025-02-14 14:01:10.157Z qt warning                                 topics="qt" tid=28733564 text="TypeError: Cannot call method 'refreshTxSettings' of null" file=qrc:/app/AppLayouts/Wallet/popups/simpleSend/SendSignModal.qml:122 category=default
WRN 2025-02-14 14:01:10.157Z qt warning                                 topics="qt" tid=28733564 text="Error: Invalid write to global property \"customBaseFee\"" file=qrc:/app/AppLayouts/Wallet/popups/simpleSend/SendSignModal.qml:256 category=default
WRN 2025-02-14 14:01:19.313Z qt warning                                 topics="qt" tid=28733564 text="TypeError: Value is null and could not be converted to an object" file=qrc:/app/mainui/SendModalHandler.qml:596 category=default
WRN 2025-02-14 14:01:19.313Z qt warning                                 topics="qt" tid=28733564 text="TypeError: Value is null and could not be converted to an object" file=qrc:/app/mainui/SendModalHandler.qml:596 category=default
WRN 2025-02-14 14:01:19.313Z qt warning                                 topics="qt" tid=28733564 text="TypeError: Value is null and could not be converted to an object" file=qrc:/app/mainui/SendModalHandler.qml:596 category=default
WRN 2025-02-14 14:01:19.313Z qt warning                                 topics="qt" tid=28733564 text="TypeError: Cannot call method 'refreshTxSettings' of null" file=qrc:/app/AppLayouts/Wallet/popups/simpleSend/SendSignModal.qml:122 category=default
WRN 2025-02-14 14:01:19.313Z qt warning                                 topics="qt" tid=28733564 text="Error: Invalid write to global property \"customBaseFee\"" file=qrc:/app/AppLayouts/Wallet/popups/simpleSend/SendSignModal.qml:256 category=default`


</p>
</details>

anastasiyaig commented:
@saledjenic can we fix this pls
```

[2025-02-17T12:38:18.042Z] make[4]: *** No rule to make target '../src/assets/png/wallet/network-notification-dot.png', needed by 'qrc_assetstmp.cpp'.  Stop.

[2025-02-17T12:38:18.042Z] make[3]: *** [CMakeFiles/Makefile2:269: CMakeFiles/big_resources_assets.dir/all] Error 2

[2025-02-17T12:38:18.043Z] make[2]: *** [CMakeFiles/Makefile2:307: CMakeFiles/StatusQ.dir/rule] Error 2

[2025-02-17T12:38:18.043Z] make[1]: *** [Makefile:235: StatusQ] Error 2

[2025-02-17T12:38:18.043Z] make: *** [Makefile:282: statusq-build] Error 2
```
as the app is not built

caybro commented on ui/StatusQ/src/StatusQ/Controls/StatusColorAnimation.qml:
Not sure it should be under `Controls`.... :thinking: wdyt @micieslak ?

micieslak replied:
In fact currently we don't have any rules regarding locations and various files should be (intuitively) in different locations. At some point we should clarify that and move files according to those agreements. For now, it's hard to say where given file should be :man_shrugging:

caybro commented on ui/StatusQ/src/StatusQ/Controls/StatusColorAnimation.qml:
```diff
@@ -0,0 +1,62 @@
+SequentialAnimation {
+    id: root
+
+    property var target: null
```
```suggestion
    required property var target
```

caybro commented on ui/StatusQ/src/StatusQ/Controls/StatusFeeOption.qml:
```diff
@@ -0,0 +1,233 @@
+    signal clicked()
+
+    font.family: Theme.baseFont.name
+    font.pixelSize: 12
```
```suggestion
    font.pixelSize: Theme.tertiaryTextFontSize
```

caybro commented on ui/StatusQ/src/StatusQ/Controls/StatusFeeOption.qml:
```diff
@@ -0,0 +1,233 @@
+        id: animatedText
+        verticalAlignment: Qt.AlignVCenter
+        font.family: root.font.family
+        font.pixelSize: root.font.pixelSize
```
Just do:
```qml
font: root.font
```

caybro replied:
That way changing e.g. `font.weight` at the toplevel will propagate here properly

caybro commented on ui/StatusQ/src/StatusQ/Controls/StatusFeeOption.qml:
```diff
@@ -0,0 +1,233 @@
+    horizontalPadding: 8
+    verticalPadding: 8
+
+    component AnimatedText: StatusBaseText {
```
Careful with these inline components; they tend to crash the `ui-tests` (due to the old Qt version on Ci :(

caybro replied:
And it does...

```
12:50:17  3/3 Test #3: QmlTests .........................***Exception: SegFault121.64 sec
```

caybro replied:
Well, it will still crash if you use the inline `component` here :D

saledjenic replied:
I had no idea that inline components may cause a crash. :(
I moved it to a stand-alone component, hope the tests will pass now.
@caybro thanks.

caybro replied:
Yeah they do due to a Qt bug; which is fixed in newer Qt versions

caybro replied:
But strangely enough, only in tests; the app itself will be fine

caybro commented on ui/StatusQ/src/StatusQ/Controls/StatusFeeOption.qml:
```diff
@@ -0,0 +1,233 @@
+        id: background
+        implicitHeight: 84
+        implicitWidth: 101
+        radius: 8
```
```suggestion
        radius: Theme.radius
```

caybro commented on ui/StatusQ/src/StatusQ/Controls/StatusFeeOption.qml:
```diff
@@ -0,0 +1,233 @@
+            return Theme.palette.statusAppLayout.backgroundColor
+        }
+
+        MouseArea {
```
A better idea is _not_ to include this functionality in `background`; if anyone overrides it, you'll never get the `clicked()` signal

caybro replied:
This needs addressing; just put it outside of contentItem/background, to the very bottom of this file

caybro commented on ui/StatusQ/src/StatusQ/Controls/StatusFeeOption.qml:
```diff
@@ -0,0 +1,233 @@
+
+        RowLayout {
+
+            Layout.preferredWidth: parent.width
```
```suggestion
            Layout.fillWidth: true
```

caybro commented on ui/StatusQ/src/StatusQ/Controls/StatusFeeOption.qml:
```diff
@@ -0,0 +1,233 @@
+
+            StatusImage {
+                id: image
+                Layout.alignment: Qt.AlignVCenter
```
Default

caybro commented on ui/StatusQ/src/StatusQ/Popups/Dialog/StatusDialogHeader.qml:
```diff
@@ -73,4 +80,29 @@ Rectangle {
         color: Theme.palette.dropShadow
     }
+
+    Rectangle {
```
Hmm, what's the intention of this? A popup on top of another popup's header?

caybro commented on ui/StatusQ/src/assets.qrc:
```diff
@@ -281,6 +281,7 @@
         <file>assets/img/icons/security.svg</file>
         <file>assets/img/icons/seed-phrase.svg</file>
         <file>assets/img/icons/send.svg</file>
+        <file>assets/img/icons/settings-advance.svg</file>
```
Looks like a duplicate

saledjenic replied:
Believe it or not, they are different. :)

caybro commented on ui/app/AppLayouts/Wallet/views/TransactionSettings.qml:
```diff
@@ -0,0 +1,495 @@
+            StatusBaseText {
+                Layout.preferredWidth: parent.width
+                text: qsTr("Transaction settings")
+                font.pixelSize: 17
```
```suggestion
                font.pixelSize: Theme.secondaryAdditionalTextSize
```

caybro commented on ui/app/AppLayouts/Wallet/views/TransactionSettings.qml:
```diff
@@ -0,0 +1,495 @@
+    signal cancelClicked()
+
+    color: Theme.palette.statusModal.backgroundColor
+    radius: 8
```
```suggestion
    radius: Theme.radius
```

caybro reviewed (commented):
Huge work :muscle: 

But since it touches the core components, got a few questions

micieslak commented on ui/StatusQ/src/StatusQ/Controls/StatusInput.qml:
```diff
@@ -96,6 +96,42 @@ Item {
         \endqml
     */
     property alias errorMessageCmp: errorMessage
+    /*!
```
Yeah, basically the direction is to decompose big, over-parametrized components into smaller, more specific ones, relying more on composition instead of parametrization. It's especially true for component like this which is heavily used but not covered by tests and having multiple parameters. After any change there it's very hard to be sure that nothing has been broken.

micieslak commented on ui/imports/utils/Utils.qml:
```diff
@@ -980,4 +980,24 @@ QtObject {
         return typeName
     }
+
+    function weiToEth(value) {
```
That's nice to have those UI-side method there. I think next we could finally cleanup all those conversions because there are various stores doing that conversion in qml, incorrectly, e.g.:
```
    function getWei2Eth(wei, decimals) {
        return wei/(10**decimals)
    }
```
in `TransactionStore.qml`

Or e.g.:
```
    function getWei2Eth(wei) {
        return globalUtilsInst.wei2Eth(wei,18)
    }
 ```
in `RootStore.qml`, `SwapStore` and others

micieslak reviewed (commented):
Looks good from my perspective, just some suggestions/questions for further consideration.

saledjenic commented on ui/StatusQ/src/StatusQ/Controls/StatusFeeOption.qml:
```diff
@@ -0,0 +1,233 @@
+        RowLayout {
+
+            Layout.preferredWidth: parent.width
+            Layout.preferredHeight: Math.max(mText.height, image.height)
```
@caybro it's about text and icon in the same line, while the icon is visible only on hover and without this line the text is moved a bit up/down when the user hovers over/out of the component.

saledjenic commented on storybook/pages/TransactionSettingsPanelPage.qml:
@micieslak I don't see the difference here?

saledjenic commented:
@micieslak @caybro regarding the location of added components, exactly that, we still don't have any rules established and I added it intuitively.

Regarding StautsInput, I agree it's a pretty complex component and used in a lot of places, and agree that we need to decompose it, but the part that I've added there should not affect any old usage of that component, did it carefully.

Thanks for the review, I've addressed your comments.

caybro reviewed (commented):
Just 2 things left for me

saledjenic commented:
@caybro should be all good now.

caybro reviewed (approved):
LGTM

But you really have to fix the ui-tests segfault :)

virginiabalducci commented:
The warning (orange) and error messages (red) when inputting different amounts on the Custom fees elements (Max base fee, Priority fee, Max gas amount, Nonce) have been tested and triggered as expected. 

There is a bug that happens when user edits the custom fees once, but then when wanting to re-edit, the 'confirm button' does not enable even when all amounts are valid. Not sure if this is related to this PR or a different known issue. Could this be related to the [issue 1](https://github.com/status-im/status-desktop/pull/17118#issuecomment-2659643404) reported here? (related to https://github.com/status-im/status-desktop/issues/17291) 

https://github.com/user-attachments/assets/cc896b7d-d5c1-4d0d-97b1-d74baf7087ab

https://github.com/user-attachments/assets/79c239fe-3f12-4939-9c03-11e3548b10ea

saledjenic commented:
@virginiabalducci I see what's the issue, should be good now, please have a look.

Regarding this...
> Also, before clicking on 'Review Send; the Estimated time on the Send modal says 'Unknown', could it be related to https://github.com/status-im/status-desktop/pull/17306 ?

That should be sorted out in the follow up PR, here https://github.com/status-im/status-desktop/pull/17331

virginiabalducci commented:
This looks good now!
